dds/idl/langmap_generator.cpp

```diff
@@ -1597,7 +1597,8 @@ struct Cxx11Generator : GeneratorBase
     const std::string lang_field_type = generator_->map_type(type);
     const Classification cls = classify(actual_field_type);
     if (!(cls & (CL_PRIMITIVE | CL_ENUM))) {
-      return "    new(&_" + name + ") " + lang_field_type + ";\n";
+      return "    new(&_" + name + ") " + lang_field_type + ";\n"
```

Comment:
The change in this file was related to a change in a test which has since been reverted.  Should this also be reverted?  I don't see a need for it, but let me know if I'm missing something.

Reply:
I would prefer it stay if we're not throwing an error for it yet. A user using OpenDDS incorrectly is still using OpenDDS and think we should be able to handle situations like this if it's not a burden on us.

Reply:
I agree with the overall goal.  I'm just not sure which kinds of misuse this protects against.

The version on master doesn't use `_set` for much at all.  It could probably do without it.

To protect against some misuse, instead of using `_disc` (along with `_set`) to determine the active member of the union, we could add a second data member of the same type.  This wouldn't be changed by `_d()`.

Reply:
So maybe something like this:

```C++
void MiscUnion::_activate(int16_t d)
{
  if (d != _ctor_disc) {
    _reset();
  }
  switch (d) {
  case 0: {
    break;
  }
  case 1:
  case 2: {
    new(&_string_value) std::string;
    break;
  }
  case 3: {
    break;
  }
  default:
    break;
  }
  _disc = d;
  _ctor_disc = d;
}

void MiscUnion::_reset()
{
  switch (_ctor_disc) {
  case 0: {
    break;
  }
  case 1:
  case 2: {
    _string_value.~basic_string();
    break;
  }
  case 3: {
    break;
  }
  default:
    break;
  }
}
```

Reply:
So maybe something like this?

```C++
void MiscUnion::_activate(int16_t d, bool init = false)
// Set init to true in constructor so reset won't act based on garbage.
{
  if (!init && d != _constructor_disc) {
    _reset();
  }
  _disc = d;
  _constructor_disc = d;
  switch (_constructor_disc) {
  case 0: {
    break;
  }
  case 1:
  case 2: {
    new(&_string_value) std::string;
    break;
  }
  case 3: {
    break;
  }
  default:
    break;
  }
}

void MiscUnion::_reset()
{
  switch (_constructor_disc) {
  case 0: {
    break;
  }
  case 1:
  case 2: {
    _string_value.~basic_string();
    break;
  }
  case 3: {
    break;
  }
  default:
    break;
  }
}
```

Reply:
Yes, this is pretty close to what I was thinking.  In looking at it again, maybe we name it `_active` or something similar.  Every switch generated (operator=, etc.) would use this.  It doesn't provide error handling in all cases (that could come later, depending on how we want to do it).

As long as were here, we could attempt to deal with potential name conflicts between user-provided names and generated names.  (While also not violating C++ name requirements.)

Reply:
Let's merge this one without the langmap_generator.cpp changes.  They are not related to multitopic / content subscription.